docs/design/proposal-troubleshoot-consolidated-cli.md

```diff
@@ -0,0 +1,350 @@
+
+This new package should be kept as minimal as possible. and serve only as an interface to private functions in the other packages.
+
+Once the stable API is ready we can instruct projects like kurl to target that and work on removing code marked for deprecation and migrate non-public functions to an `internal` package to make it clear that it's not intended to be imported.
+
+The core of this api will be exposed via new troubleshoot `types`
+
+```go
+type TroubleshootKinds struct {
+}
+
+type Bundle struct {
+  FilePath string
+}
+
+type AnalysisResults struct {
+  Bundle Bundle // include bundle metadata
+}
+
+```
+
+from there we can build out methods for each to interact with them:
+
+```go
+// Load loads specs defined by the LoadOptions struct and returns a TroubleshootKinds struct
+func Load(LoadOptions) (TroubleshootKinds,err){}
+
+// Bundle.Collect runs collections defined in TroubleshootKinds passed through CollectOptions
+func (bundle *Bundle) Collect(CollectOptions) (error) {}
+
+// Bundle.Load will create a Bundle struct from options in LoadBundleOptions
+func (bundle *Bundle) Load(LoadBundleOptions) (error) {}
+// it's worth noting that while this may appear inefficient now
+// it allows us to extend what we include in the Bundle struct in future without having to continuously extend the load function.
+
+// Bundle.Analyze runs analysis defined in TroubleshootKinds passed through AnalyzeOptions
+func (bundle *Bundle) Analyze(AnalyzeOptions) (AnalysisResults, error) {}
+
+// Bundle.Redact runs redaction defined in TroubleshootKinds passed through RedactOptions
+func (bundle *Bundle) Redact(RedactOptions) error {}
+
+// Bundle.Archive produces an archive from a bundle on disk with options passed in ArchiveOptions
+func (bundle *Bundle) Archive(ArchiveOptions) error {}
+
+// Bundle.Serve starts an sbctl-like server with options defined in ServeOptions
+func (bundle *Bundle) Serve(ServeOptions) error {}
+
+// AnalysisResults.Save saves the results of the analysis according to SaveAnalysisOptions
+func (results *AnalysisResults) Save(SaveAnalysisOptions) error {}
+```
+
+To Allow for forward flexibility while reducing the impact of potentially breaking changes, options will be passed into these methods via options structs.
+
+```go
+type LoadOptions struct {
+  RawSpecs []string // list of locations for specs to load
+  RawSpec string // a single spec to parse
+  FilePaths []string // list of filepaths to check, can be globbed
+  URIs []string // list of URIs to retrieve specs from
+  SearchCluster bool // toggle for searching cluster from context for troubleshoot objects
+  ProgressChan chan // a channel to write progress information to
+}
+
+type LoadBundleOptions struct {
+    Path string // Path to archive or directory of bundle files
+}
+
+type CollectOptions struct {
+  DefaultCollectors []Collector // list of collectors to run regardless of what's in the spec
```

Comment:
With the default collectors/redactors/analysers, I think we should not treat them in any special way. I see it working like so

```go
defaults := TroubleshootKinds{}
defaults.CollectorsV1Beta2 = append(kinds.CollectorsV1Beta2, other.CollectorsV1Beta2...)

kinds := TroubleshootKinds{}
kinds.Add(defaults)

opt := CollectOptions{Specs: kinds}
```

Reply:
Troubleshoot library will expose a `specs.DefaultSpecs` variable somewhere which `support-bundle` & `preflight` binary would consume.